### PR TITLE
feat: migrate deepseek models to nnx

### PR DESCRIPTION
**Pull request by mesakhcienet**
# Description

This PR migrates the `deepseek` and `deepseek_batch_split` models to use the NNX framework.

Additionally, `layerwise_quantizations.py` has been updated to use the `ToLinen` version of the DeepSeek layer.

**Background:**

*   **Re-landing Previous Work:** This PR re-introduces changes from two previously reviewed Pull Requests:
    *   PR #2098 
    *   PR #2522 
    The core logic in this PR has already been reviewed in those previous iterations.

*   **Anticipated Internal Test Failures:** We anticipate potential failures in Google3 presubmit tests when this change is imported internally. These are expected **due to differences between the external and internal testing environments and dependencies**. We are prepared to address and fix any internal test failures that arise during the internal review process.

# Tests

We use [xpk](https://github.com/AI-Hypercomputer/xpk) to create tpu cluster and assign workload

## Cluster
TPU type : v6e-32
Number of slices : 4
GKE version : 1.31.11-gke.1036000
Base Image : us-docker.pkg.dev/cloud-tpu-images/jax-ai-image/tpu:jax0.7.0-rev1 

## Image

Build Image command :
```bash
bash docker_build_dependency_image.sh DEVICE=tpu MODE=stable_stack BASEIMAGE=us-docker.pkg.dev/cloud-tpu-images/jax-ai-image/tpu:jax0.7.0-rev1
```

### Original `Deepseek` Run Xpk command :
  
  ``` bash
  python xpk.py workload create --cluster $CLUSTER_NAME \
   --base-docker-image mesa_maxtext_base_image \
   --workload=$WORKLOAD \
   --tpu-type=${TPU_TYPE} --num-slices=${NUM_SLICES} --max-restarts=10 \
     --on-demand \
    --script-dir=$MAXTEXT_SCRIPT_DIR --command \
     "python3 -m MaxText.train MaxText/configs/base.yml \
     run_name=runner_direct_${idx}  \
     base_output_directory=${BASE_OUTPUT_DIRECTORY}  \
     model_name=${MODEL_NAME} \
     dataset_type=synthetic  \
     async_checkpointing=false  \
     per_device_batch_size=1 \
     metrics_file='metrics.txt'  \
     steps=15"
  ```
### `Deepseek_batch_split` Run Xpk command :

  ``` bash
  python xpk.py workload create --cluster $CLUSTER_NAME \
   --base-docker-image mesa_maxtext_base_image \
   --workload=$WORKLOAD \
   --tpu-type=${TPU_TYPE} --num-slices=${NUM_SLICES} --max-restarts=10 \
     --on-demand \
    --script-dir=$MAXTEXT_SCRIPT_DIR --command \
     "python3 -m MaxText.train MaxText/configs/base.yml \
     run_name=runner_direct_${idx}  \
     base_output_directory=${BASE_OUTPUT_DIRECTORY}  \
     model_name=${MODEL_NAME} \
     dataset_type=synthetic  \
     async_checkpointing=false  \
     per_device_batch_size=1 \
     metrics_file='metrics.txt'  \
     steps=15 \
     use_batch_split_schedule=true" 
  ```
  
  ### layerwise quantization Run command:
  
 
```bash
LOAD_PARAMS_PATH=/dev/shm/deepseek2-16-chat-unscanned/unscanned/
SAVE_PARAMS_PATH=/dev/shm/deepseek2-16-chat-unscanned-linen/
python -m MaxText.layerwise_quantization src/MaxText/configs/base.yml \
  tokenizer_path=src/MaxText/assets/tokenizer.llama2 load_parameters_path=${LOAD_PARAMS_PATH} \
  model_name=deepseek2-16b \
  ici_tensor_parallelism=-1 scan_layers=false weight_dtype=bfloat16 per_device_batch_size=1 \
  attention=dot_product quantization=int8 ici_fsdp_parallelism=8 async_checkpointing=false enable_single_controller=true \
  tokenizer_type=huggingface megablox=false sparse_matmul=false \
  checkpoint_storage_use_ocdbt=False \
  checkpoint_storage_use_zarr3=False \
  save_quantized_params_path=${SAVE_PARAMS_PATH}  

```
  

### Original Deepseek
- Before migration (from main branch) : [link](https://paste.googleplex.com/6636920108744704#)
- After migration (train from scratch] : [link](https://paste.googleplex.com/5134325188263936#) 
- After migration (train with loading previous model from main branch / before migration, with `steps` argument sets from `15` to `50` ): [link](https://paste.googleplex.com/6589488327426048#)

- Profiling : https://github.com/AI-Hypercomputer/maxtext/pull/2098#issuecomment-3383398059

### Deepseek batch split

Train diff before and after migration [here](https://diff.googleplex.com/#key=4EtN9XxcFDga)

- Before migration (from main branch) : [link](https://paste.googleplex.com/6384283723890688)
- After migration (train from scratch) : [link](https://paste.googleplex.com/6281821843226624)
- After migration (train with loading previous model from main branch / before migration, with `steps` argument sets from `15` to `40` ): [link](https://paste.googleplex.com/5412037077368832)

- Profiling : https://github.com/AI-Hypercomputer/maxtext/pull/2522#issuecomment-3444149597

### Layerwise quantization

Before changes (`main` branch)
- Missing arguments error log: https://paste.googleplex.com/6276009152348160
- TypeError log: https://paste.googleplex.com/6441677845823488#l=549
- ValueError log: https://paste.googleplex.com/6355183384788992

After changes (current fix):
- Quantization Process Log: gs://mesa-maxtext-bucket/layerwise_quantization_log.txt
- Quantized Checkpoint Structure: 
    - Before (main branch after code fixed) : https://paste.googleplex.com/6238853356716032
    - After  (current PR)   : https://paste.googleplex.com/6245295321120768
- Script to show structure: https://paste.googleplex.com/4911461602295808
- Comparison of weight dtype, shape and value : https://paste.googleplex.com/4878258770411520  
  Script : https://paste.googleplex.com/5482582965878784 
 Note: we use absolute and relative tolerance to `1e-3`, because we compare linen and ported to nnx model which have different random initializer 

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/layerwise_quantization.py [hunk 1]
```diff
@@ -108,33 +213,50 @@ def model_apply(_p, _rng, layer):
...
+        aqt_vars = new_vars["aqt"]
+
+        try:
+          removed_params = remove_quantized_params(params["params"], aqt_vars)
+          quantized_params["params"]["decoder"][layer_name] = removed_params
+        except Exception as e:
+          print(f"ERROR: Failed to remove quantized params for {layer_name}: {e}")
+          print(f"DEBUG: Dumping params['params'] keys for {layer_name}:")
+          jax.tree_util.tree_map_with_path(lambda path, _: print(f"  {jax.tree_util.keystr(path)}"), params["params"])
+          print(f"DEBUG: Dumping new_vars['aqt'] keys for {layer_name}:")
```
Comment (NicoGrande) [inline src/MaxText/layerwise_quantization.py]: Can we replace the print statements with max_logging?

File: src/MaxText/layerwise_quantization.py [hunk 2]
```diff
@@ -144,6 +266,8 @@ def _load_layer(self, layer_name):
 
     config = self.config
     with nn_partitioning.axis_rules(config.logical_axis_rules):
+
+      # print(f"DEBUG: Abstract params for {layer_name}: {jax.tree_map(lambda x: x.shape, partial_abstract_params)}")
```
Comment (NicoGrande) [inline src/MaxText/layerwise_quantization.py]: We can remove this commented line if it is not needed.

File: src/MaxText/layerwise_quantization.py [hunk 3]
```diff
@@ -108,33 +213,50 @@ def model_apply(_p, _rng, layer):
...
           mutable=True,
       )
 
     for layer, num_layers, layer_prefix in zip(layers, num_layers_list, layer_prefixes):
       for index in tqdm(range(num_layers)):
-        print(f"Quantizing layer {layer_prefix}_{index}")
-
         layer_name = f"{layer_prefix}_{index}"
-        params = self._load_layer(layer_name)
+        print(f"Processing layer: {layer_name}")
```
Comment (NicoGrande) [inline src/MaxText/layerwise_quantization.py]: Same here about max_logging.

File: src/MaxText/layerwise_quantization.py [hunk 4]
```diff
@@ -154,46 +278,38 @@ def _load_layer(self, layer_name):
...
-      return IGNORE
+      if not _should_keep(path, value):
+        return IGNORE
+      if isinstance(value, jax.ShapeDtypeStruct):
+        zeros_array = jnp.zeros(value.shape, value.dtype)
+        if value.sharding is not None:
+          try:
+            return jax.device_put(zeros_array, value.sharding)
+          except Exception as e:  # pylint: disable=broad-except
+            print(f"Error applying sharding for path {path}: {e}")
```
Comment (NicoGrande) [inline src/MaxText/layerwise_quantization.py]: One more print statement here.

File: src/MaxText/layerwise_quantization.py [hunk 5]
```diff
@@ -77,29 +178,33 @@ def __init__(self, config: Any):
...
-        deepseek.DeepSeekMoELayer(config, mesh=self._mesh, quant=self.quant),
+        deepseek.DeepSeekMoELayerToLinen(
+            config=config, mesh=self._mesh, quant=self.quant, model_mode=model_mode, rngs=nnx.Rngs(self.rng)
+        ),
+        deepseek.DeepSeekDenseLayerToLinen(
+            config=config, mesh=self._mesh, quant=self.quant, model_mode=model_mode, rngs=nnx.Rngs(self.rng)
+        ),
     ]
-    layer_prefixes = ["dense_layers", "moe_layers"]
+    layer_prefixes = ["moe_layers", "dense_layers"]
```
Comment (RissyRan) [inline src/MaxText/layerwise_quantization.py]: Is this opposite order on purpose? DeepSeek has dense first and then moe layers

Comment (mesakhcienet) [reply src/MaxText/layerwise_quantization.py]: it was on debug purpose. I will revert it back to follow the earlier order, thank you.

for the `test_deepseek_quantize.sh` script , we use the command below :

```bash
LOAD_PARAMS_PATH=/dev/shm/deepseek2-16-chat-unscanned/unscanned/
SAVE_PARAMS_PATH=/dev/shm/deepseek2-16-chat-unscanned-linen/
python -m MaxText.layerwise_quantization src/MaxText/configs/base.yml \
  tokenizer_path=src/MaxText/assets/tokenizer.llama2 load_parameters_path=${LOAD_PARAMS_PATH} \
  model_name=deepseek2-16b \
  ici_tensor_parallelism=-1 scan_layers=false weight_dtype=bfloat16 per_device_batch_size=1 \
  attention=dot_product quantization=int8 ici_fsdp_parallelism=8 async_checkpointing=false enable_single_controller=true \
  tokenizer_type=huggingface megablox=false sparse_matmul=false \
  checkpoint_storage_use_ocdbt=False \
  checkpoint_storage_use_zarr3=False \
  save_quantized_params_path=${SAVE_PARAMS_PATH}  

```

File: src/MaxText/layerwise_quantization.py (hunk 1 above)
Comment (mesakhcienet) [reply src/MaxText/layerwise_quantization.py]: sure, and done. Thank you

**Comment by mesakhcienet** [comment]
These changes have been integrated into our internal codebase. This was done via Google's internal tooling and submitted as [cl/852094911](https://cl.corp.google.com/852094911) (note: this is an internal Google link). We'll be closing this PR.
